drivers: pse: si3474: new pse driver for si3474

robi251 opened the pull request:
New driver of the Si3474 (Silicon Laboratories) Power Source Equipment functionalities:
- reading voltages for particular PSE channels
- reading current for particular PSE channels
- reading main voltage of the PSE device
- reading chip temperature
- reading all events in runtime

Commit:
drivers: pse: si3474: new pse driver for si3474

New driver of the Si3474 (Silicon Laboratories) Power Source
Equipment functionalities:
- reading voltages for particular PSE channels
- reading current for particular PSE channels
- reading main voltage of the PSE device
- reading chip temperature
- reading all events in runtime

Signed-off-by: Robert Slawinski <robert.slawinski.ext@indurad.com>

jeppenodgaard commented on drivers/pse/si3474/si3474.c:
```diff
@@ -0,0 +1,383 @@
+	res = i2c_reg_write_byte(cfg->i2c.bus, dev_address, INTERRUPT_MASK, events);
+	if (res != 0) {
+		LOG_WRN("Si3474 setting events failed!");
+		return res;
```
Can be removed

robi251 replied:
Maybe I should rename it to activate events? si3474_set_events() is configuring which types/events will generates the interrupt. Generated events are fetched from the sensor by si3474_get_events();

jeppenodgaard replied:
I just meant that the `return` could be removed because there's a return right after.  I think the function name is fine.

jeppenodgaard commented on drivers/pse/si3474/si3474.c:
```diff
@@ -0,0 +1,383 @@
+
+	switch (channel) {
+	case CHANNEL_0:
+		res = i2c_burst_read(cfg->i2c.bus, dev_address, PORT1_CURRENT_LSB_REG, &buff[0],
```
This is the only line that seem to change from each case. Is it possible to store the define in a variable and move the redundant code outside the switch case?
The voltage function seem to do pretty much the same. Maybe write a helper function.

robi251 replied:
Yes, correct, the redundant code can be moved outside of the switch to reduce the source code. This will be fixed.

jeppenodgaard commented on drivers/pse/si3474/si3474.c:
```diff
@@ -0,0 +1,383 @@
+	res = i2c_reg_write_byte(cfg->i2c.bus, dev_address, PB_POWER_ENABLE_REG,
+				 (1 << (channel + 4)));
+	if (res != 0) {
+		LOG_WRN("Si3474 switching channel failed");
```
Maybe print channel parameter

robi251 replied:
Agree - this can give more information about potential fail. Will be fixed.

jeppenodgaard commented on drivers/pse/si3474/si3474.c:
```diff
@@ -0,0 +1,383 @@
+
+	res = si3474_init_ports(dev);
+	if (res != 0) {
+		LOG_ERR("Configuring Si3474 interrupt and oss pins failed");
```
I dont know what oss is 😅 If it is an abbreviation maybe it should be capital letters.
Edit: Ah, over supply signal. Maybe it's ok since it's in the binding

robi251 replied:
I will correct description to Over Supply Signal ;)

jeppenodgaard commented on drivers/pse/si3474/si3474_setup.c:
```diff
@@ -0,0 +1,143 @@
+	uint8_t buff;
+	int res;
+
+	if (!device_is_ready(cfg->gpio_rst.port)) {
```
There is a gpio is ready function. Cannot remember the exact name.

robi251 replied:
Agree - I will fix it.

jeppenodgaard commented on drivers/pse/si3474/si3474_setup.c:
```diff
@@ -0,0 +1,143 @@
+		return res;
+	}
+
+	k_msleep(1000);
```
I think this will delay the start time of all threads. Does i really need to wait a second? Can it be solved another way?

robi251 replied:
This is typical copy-paste error :| Here must be 10ms k_msleep(10);. Will be corrected.

jeppenodgaard commented on include/zephyr/drivers/pse.h:
```diff
@@ -0,0 +1,420 @@
+ */
+enum pse_mode {
+	PSE_DEACTIVATED,
+
```
Ekstra line

robi251 replied:
Sure - Will be corrected

jeppenodgaard reviewed (commented):
Nice work! Some nit comments.

marcusfolkesson commented on drivers/pse/si3474/si3474.c:
```diff
@@ -0,0 +1,383 @@
+	res = si3474_i2c_read_reg(cfg->i2c.bus, dev_address, INTERRUPT_REGISTER,
+				  &pse_interrupt_buffer);
+	if (res != 0) {
+		LOG_WRN("Si3474 reading interrupt status failed");
```
Do you still want to continue if you failed to read the interrupt status?

robi251 replied:
For sure - must be corretced.

marcusfolkesson commented on drivers/pse/si3474/si3474.c:
```diff
@@ -0,0 +1,383 @@
+		switch (pse_interrupt_buffer & mask) {
+		case POWER_ENABLE_CHANGE_IT:
+			mask <<= 1;
+			cnt++;
```
Forgot a break?
If not, please add __fallthrough;

robi251 replied:
__fallthrough; must be added. break is missed intentionally. Will be corrected.

marcusfolkesson commented on drivers/pse/si3474/si3474.c:
```diff
@@ -0,0 +1,383 @@
+
+		case DETECT_CC_DONE_IT:
+			mask <<= 1;
+			cnt++;
```
break or __fallthrough;

robi251 replied:
__fallthrough;  ;)

robi251 commented on drivers/pse/si3474/si3474.c:
```diff
@@ -0,0 +1,383 @@
+	int res = i2c_reg_write_byte(i2c, dev_addr, reg_ddr, val);
+	if (res != 0) {
+		LOG_WRN("Si3474 device 0x%x writing 0x%x failed! [%i]", dev_addr, reg_ddr, res);
+		return res;
```
Agree - will be removed.

robi251 commented on drivers/pse/si3474/si3474.c:
```diff
@@ -0,0 +1,383 @@
+	res = i2c_reg_read_byte(i2c, dev_addr, reg_ddr, buff);
+	if (res != 0) {
+		LOG_WRN("Si3474 device 0x%x reading 0x%x failed...! [%i]", dev_addr, reg_ddr, res);
+		return res;
```
Agree - will be removed.

Commit:
pse: corrections after codereview

Most of corrections incorporated after test on the real hardware.
Some of them, which are still under discussion, are waiting for
decision.

Signed-off-by: Robert Slawinski <robert.slawinski.ext@indurad.com>

jeppenodgaard commented on drivers/pse/si3474/si3474.c:
```diff
@@ -344,7 +301,7 @@ static int si3474_init(const struct device *dev)
 	res = si3474_init_ports(dev);
 	if (res != 0) {
-		LOG_ERR("Configuring Si3474 interrupt and oss pins failed");
+		LOG_ERR("Configuring Si3474 interrupt and Over Suply Source pin failed");
```
Supply

jeppenodgaard commented on drivers/pse/si3474/si3474.c:
```diff
@@ -51,87 +41,90 @@ static int si3474_get_events(const struct device *dev, uint8_t *events)
-						  &events[DISCONNECT_PCUT_FAULT]);
-			if (res != 0) {
-				LOG_WRN("Si3474 reading event status table failed");
+	}else {
```
Missing white space.

Commit:
pse: corrections after code review

Somme clean code corrections has been made after code review

Signed-off-by: Robert Slawinski <robert.slawinski.ext@indurad.com>

robi251 reviewed (commented):
All findings corrected

jeppenodgaard reviewed (approved):
LGTM!

robi251 commented:
> LGTM!

OK, then we can merge it I think?

BTW. During the development I have create more functionalities for this driver, but I don't want to make a mess here. I will create separate Pull-Request.

marcusfolkesson reviewed (approved):
LGTM

fabiobaltieri commented:
Hi @robi251, new API introduction should have an [RFC](https://github.com/zephyrproject-rtos/zephyr/issues/new?assignees=&labels=RFC&projects=&template=003_rfc-proposal.md&title=) have to go through the Architecture Review process, see the documentation at: https://docs.zephyrproject.org/latest/contribute/proposals_and_rfcs.html

Basically it would be about justifying the creation of the new "PSE" subsystem and how it abstract a class of devices. Could you go through that process with this one?

marcusfolkesson commented on drivers/pse/si3474/si3474_setup.c:
```diff
@@ -0,0 +1,141 @@
+{
+	struct si3474_data *drv_data = dev->data;
+	const struct si3474_config *cfg = dev->config;
+	int res = 0;
```
Remove res and just return 0

marcusfolkesson commented on drivers/pse/si3474/si3474_reg.h:
```diff
@@ -0,0 +1,63 @@
+#define DETECT_CC_DONE_IT      0x8
+#define CLASS_DONE_IT          0x10
+#define P_I_FAULT_IT           0x20
+#define START_EVENT_IT         0x64
```
This is wrong... it should be 0x40 according to datasheet.

However, please use BIT() macro for this.

#define POWER_ENABLE_CHANGE_IT   BIT(0)
#define POWER_GOOD_CHANGE_IT   BIT(1)
...

marcusfolkesson commented on drivers/pse/si3474/si3474_setup.c:
```diff
@@ -0,0 +1,141 @@
+{
+	struct si3474_data *drv_data = dev->data;
+	const struct si3474_config *cfg = dev->config;
+	int result = 0;
```
_res_ is used all over the place to carry the result. Please be consistent.
Rename result to res.

robi251 commented:
> Hi @robi251, new API introduction should have an [RFC](https://github.com/zephyrproject-rtos/zephyr/issues/new?assignees=&labels=RFC&projects=&template=003_rfc-proposal.md&title=) have to go through the Architecture Review process, see the documentation at: https://docs.zephyrproject.org/latest/contribute/proposals_and_rfcs.html
> 
> Basically it would be about justifying the creation of the new "PSE" subsystem and how it abstract a class of devices. Could you go through that process with this one?
> 
> You should also consider fitting this into existing subsystem, maybe a mix of regulator for the power supply channels + sensor for the temperature and maybe other stuff, and ultimately you can always add some extra device specific API if that's unique for this device. At first sight the API seems a bit too speciic.

Yes I agree - since this driver is changing the Zephyr API I need to prepare the RFC. Hopefully I will do this today.

kartben commented:
Can't this simply use the regulator API?

aasinclair commented:
@kartben Looking at the datasheet I agree.

This looks like a good fit for mfd + regulators + watchdog etc.
Existing PMIC drivers in the tree are implemented like this.

@robi251 Please take a look at the existing drivers for ad5592, max20335 and npm1300

MaureenHelm commented:
> @kartben Looking at the datasheet I agree.
> 
> This looks like a good fit for mfd + regulators + watchdog etc. Existing PMIC drivers in the tree are implemented like this.
> 
> @robi251 Please take a look at the existing drivers for ad5592, max20335 and npm1300

@robi251 ping

robi251 commented:
> > @kartben Looking at the datasheet I agree.
> > This looks like a good fit for mfd + regulators + watchdog etc. Existing PMIC drivers in the tree are implemented like 

> > @kartben Looking at the datasheet I agree.
> > This looks like a good fit for mfd + regulators + watchdog etc. Existing PMIC drivers in the tree are implemented like this.
> > @robi251 Please take a look at the existing drivers for ad5592, max20335 and npm1300
> 
> @robi251 ping

You are right, that some of the functionalities like:
- setting current limit
- set mode
- enable/disable
- get current/voltage
can stick to PSE driver.

However - Si3474 PSE chip has also functionalities, which are defined by the IEEE 802.3af, 802.3at, and 802.3bt standards like:
- connection detection (auto an manual)
- classification functionality (auto and manual)
- defining group of devices with supply priority
- activating/deactivating PoE ports
- generating set of events described situations like (overcurrent, overvoltage, PoE cable connection/disconnection, power classification and connection detection, overtemperature, quad configuration 4PP and 2PP etc.)

I'm not sure if this will be compatible/reasonable to use regulators API, since standard regulators are not designed to fulfill IEEE standard. Moreover - new PSE-like chips appear on the market, compliant with this standard and having almost the same (or event more sophisticated) functionalities (Analog Devices: LTC9101-2A , LTC4296-1, MAX5980GTJ+T; Texas Instruments: TPS2386). In my opinion treating them as the same as voltage regulators may be somewhat misleading.

fabiobaltieri commented:
> I'm not sure if this will be compatible/reasonable to use regulators API

I would still put it under regulator and use device specific APIs for the rest, there's plenty of examples on that (just look under `include/zephyr/drivers/sensor`), @gmarull any opinions?

aasinclair commented:
> > > @kartben Looking at the datasheet I agree.
> > > This looks like a good fit for mfd + regulators + watchdog etc. Existing PMIC drivers in the tree are implemented like
> 
> > > @kartben Looking at the datasheet I agree.
> > > This looks like a good fit for mfd + regulators + watchdog etc. Existing PMIC drivers in the tree are implemented like this.
> > > @robi251 Please take a look at the existing drivers for ad5592, max20335 and npm1300
> > 
> > 
> > @robi251 ping
> 
> You are right, that some of the functionalities like:
> 
> * setting current limit
> * set mode
> * enable/disable
> * get current/voltage
>   can stick to PSE driver.
> 
> However - Si3474 PSE chip has also functionalities, which are defined by the IEEE 802.3af, 802.3at, and 802.3bt standards like:
> 
> * connection detection (auto an manual)
> * classification functionality (auto and manual)
> * defining group of devices with supply priority
> * activating/deactivating PoE ports
> * generating set of events described situations like (overcurrent, overvoltage, PoE cable connection/disconnection, power classification and connection detection, overtemperature, quad configuration 4PP and 2PP etc.)
> 
> I'm not sure if this will be compatible/reasonable to use regulators API, since standard regulators are not designed to fulfill IEEE standard. Moreover - new PSE-like chips appear on the market, compliant with this standard and having almost the same (or event more sophisticated) functionalities (Analog Devices: LTC9101-2A , LTC4296-1, MAX5980GTJ+T; Texas Instruments: TPS2386). In my opinion treating them as the same as voltage regulators may be somewhat misleading.

You can't fit all the features in a regulator driver, but with an MFD device, you can combine multiple APIs.

For example:
* regulator specific: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/regulator/regulator_npm1300.c
* connection / disconnection events: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/mfd/mfd_npm1300.c
* gpio: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/gpio/gpio_npm1300.c
* Input reading / attributes: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/npm1300_charger/npm1300_charger.c

robi251 commented:
Sorry for delay. I was busy because of other tasks. So in principle you suggesting to:

1. Create si3474 mfd device and put all stuff relates to the connection detection, classification and events handling in:  `drivers/mfd/si3574.c` and then create the `include/zephyr/drivers/mfd/mfd_si3474.h` API which will use it.
2. Create si3474 regulator device in: `drivers/regulator/regulator_si3474.c` and current existing API `include/zephyr/drivers/regulator.h` will use it.
3. Create the si3474 gpio device in: `drivers/gpio/gpio_si3474.c` and current existing API `include/zephyr/drivers/gpio.h` will use it.
4. Create the si3474 sensor device in: `drivers/sensor/skyworks/si3474/si3474.c` and use currently existing sensor API to read parameters like si3474 case temperature.

Am I understand correctly? If yes, my only concern is that at some point there will be too many different mfds in the project.

robi251 commented:
BTW - can we reopen this PR?
